Improve performance of Group-Object

powercode opened the pull request:
Fixes #7409.

## PR Summary

code cleanup
object[] -> IList<T> in PSType.ArrayToTuple signature to reduce allocations.
Sort input before grouping. Use sorting to only consider last group when comparing.
adding DebuggerDisplay for GroupInfo.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

Commit:
Initial cleanup.

Refactoring only - no functional changes.

Commit:
object[] -> IList<T> in PSType.ArrayToTuple signature

This reducing alloctions and object copying.

Commit:
Adding DebuggerDisplay for GroupInfo

Commit:
Sort input before grouping. Use sorting to only consider last group when comparing.

Much faster when the number of unique values grows large.

Commit:
Fixing DebuggerDisplay for CommandInfo.

Commit:
Addressing code-factor issues

powercode committed:
Using a stable sort algorithm so that Group order is maintained.

powercode committed:
More code factor fixes

powercode commented:
Appveyor error seem unrelated to my changes.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -113,10 +116,10 @@ private static string BuildName(List<ObjectCommandPropertyValue> propValues)
             {
-                if (propValue != null && propValue.PropertyValue != null)
+                var propValuePropertyValue = propValue?.PropertyValue;
+                if (propValuePropertyValue != null)
```
Is the check really needed?

powercode replied:
yes, it is dereferenced in the else clause.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -246,53 +248,54 @@ public SwitchParameter NoElement
             OrderByPropertyComparer orderByPropertyComparer)
         {
-            if (currentObjectEntry != null && currentObjectEntry.orderValues != null && currentObjectEntry.orderValues.Count > 0)
+            var currentObjectorderValues = currentObjectEntry.orderValues;
```
Should it be `currentObjectOrderValues`?

powercode replied:
yes

iSazonov replied:
I don't see the fix.

iSazonov replied:
Please address the comment and fix case.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -246,53 +248,54 @@ public SwitchParameter NoElement
-                        //add this inputObject to an existing group
-                        currentGroupInfo.Add(currentObjectEntry.inputObject);
-                    }
+                    //add this inputObject to an existing group
```
Please add space after //.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -336,25 +341,31 @@ protected override void ProcessRecord()
-                        foreach (GroupInfo _grp in _groups)
+                        foreach (GroupInfo grp in _groups)
                         {
                             if (AsString)
```
Make sense to move the cycle into the  `if (AsString)` ?

powercode replied:
Not sure i understand what you suggest here.

powercode replied:
Yes, I think that is clearer.

iSazonov replied:
I thought more about performancethough :-) From this point of view it seems `for()` better than `foreach` - less allocations. Is the _groups so much that we can get a win?

powercode committed:
Addressing review comments

Also some additional CodeFactor complaints

powercode commented:
I would like to fix the parameter sets too. 
```
[Parameter(ParameterSetName = "HashTable", Mandatory = true)]
public SwitchParameter AsHashTable { get; set; }

[Parameter(ParameterSetName = "HashTable")]
public SwitchParameter AsString { get; set; }
```

powercode committed:
Refactoring Group-Object logic.

Moving parameter checking earlier.

Adding back the slow quadratic code path for the cases where we have more than one set of input types.

powercode committed:
Fixing error with determining if there are multiple types involved.

iSazonov commented:
I believe we shouldn't mess performance changes with other and it is better fix parameter sets in new PR.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -302,22 +359,41 @@ protected override void ProcessRecord()
+                    if (AsHashTable
+                        && !AsString
+                        && (Property != null && Property.Length > 1
+                            || _orderByProperty.MshParameterList.Count > 1))
```
We'd move this on previous line.

powercode committed:
Merge branch 'master' into groupobj

powercode committed:
Reverting ParametersetName changes.

powercode committed:
CodeFactor fixes.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -83,7 +113,7 @@ internal static object ArrayToTuple<T>(IList<T> inputObjects, int startIndex)
     {
         internal GroupInfoNoElement(OrderByPropertyEntry groupValue)
-            : base(groupValue)
+        : base(groupValue)
```
Please move to previous line 115.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -224,6 +250,7 @@ public class GroupObjectCommand : ObjectBase
         /// <summary>
         /// Gets or sets the AsString parameter.
         /// </summary>
+        /// <value></value>
```
Please remove the empty XML comment.

iSazonov reviewed (commented):
Leave a comment

powercode committed:
Addressing review comments

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -131,15 +165,16 @@ private static string BuildName(List<ObjectCommandPropertyValue> propValues)
                     else
                     {
-                        sb.Append(string.Format(CultureInfo.InvariantCulture, "{0}, ", propValue.PropertyValue.ToString()));
+                        sb.Append(string.Format(CultureInfo.InvariantCulture, "{0}, ", propValuePropertyValue.ToString()));
```
Seems string.Format was optimized in .Net Core 2 but we could still use `AppendFormat()` here and in line 160 to get more readable code and exclude extra method calls.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -238,25 +269,26 @@ public SwitchParameter NoElement
             OrderByPropertyComparer orderByPropertyComparer)
         {
-            if (currentObjectEntry != null && currentObjectEntry.orderValues != null && currentObjectEntry.orderValues.Count > 0)
+            var currentObjectorderValues = currentObjectEntry.orderValues;
```
`currentObjectorderValues` should be `currentObjectOrderValues`.

iSazonov reviewed (approved):
LGTM with two minor comments.

daxian-dbw committed:
Add new line at the end of file

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -238,25 +269,26 @@ public SwitchParameter NoElement
 
-                GroupInfo currentGroupInfo = null;
-                if (groupInfoDictionary.TryGetValue(currentTupleObject, out currentGroupInfo))
+                if (groupInfoDictionary.TryGetValue(currentTupleObject, out var currentGroupInfo))
```
What's the use case for currentGroupInfo being null in the groupInfoDictionary?  If it is expected, a comment is needed; otherwise, it looks like currentObjectEntry.inputObject is dropped on the floor.

powercode replied:
I don't think that is a case that will ever occur. Just defensive programming.

Both cases are behaving as the original implementation. 

On could argue for removing the null check as it is a condition that shouldn't occur, and if it does, it is a bug that should be fixed.

iSazonov replied:
>  if it does, it is a bug that should be fixed.

So should we have runtime check and throw?

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -286,12 +318,71 @@ public SwitchParameter NoElement
+                if (groupInfoDictionary.TryGetValue(currentTupleObject, out var currentGroupInfo))
+                {
+                    // add this inputObject to an existing group
+                    currentGroupInfo?.Add(currentObjectEntry.inputObject);
```
same as above, when is currentTobpleObject present but currentGroupInfo is null.

powercode replied:
Yeah, I'm removing the null checks.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -22,28 +24,30 @@ internal static class PSTuple
         {
             Diagnostics.Assert(inputObjects != null, "inputObjects is null");
-            Diagnostics.Assert(inputObjects.Length > 0, "inputObjects is empty");
+            Diagnostics.Assert(inputObjects.Count > 0, "inputObjects is empty");
```
This is a fairly common pattern in the engine, that there are asserts, but no backing runtime checks. I have always wondered about it...

powercode replied:
I guess the thinking is that since it's an internal class, the usage can be demonstrated to be safe and the error checking skipped.

dantraMSFT replied:
Since we don't have an official policy around Assert, I'll not hold up the PR for it and you can ignore the rest. :)

I've been involved in three different threads recently where Dbg.Assert in internal code is hit in debug builds.  In two cases, crashes were occurring the release build and the third the lack of the check was masking a bug in another component.

IMHO: Since we don't test debug builds, Dbg.Assert is a placebo at best without a release check.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs:
```diff
@@ -286,12 +318,71 @@ public SwitchParameter NoElement
+        /// <param name="groups">List containing Groups.</param>
+        /// <param name="groupInfoDictionary">Dictionary used to keep track of the groups with hash of the property values being the key.</param>
+        /// <param name="orderByPropertyComparer">The Comparer to be used while comparing to check if new group has to be created.</param>
+        internal static void DoOrderedGrouping(
```
It was internal, and I just kept it so, but I'm happy to make it private.

powercode committed:
Addressing review comments from dantraMSFT and iSazonov.

powercode committed:
Addressing review comments.

Removing unnecessary call to .ToArray().

Making internal methods private.

daxian-dbw commented:
@dantraMSFT Can you please take another look?

dantraMSFT reviewed (approved):
LGTM

iSazonov commented:
@powercode Many thanks for the great perf improvement!
